src/ProductManifold.jl

```diff
@@ -0,0 +1,416 @@
+"""
+struct ProductTVector{TM<:Tuple} <: TVector
+    parts::TM
+end
+
+ProductTVector(vectors...) = ProductTVector{typeof(vectors)}(vectors)
+eltype(x::ProductTVector) = eltype(Tuple{map(eltype, x.parts)...})
+similar(x::ProductTVector) = ProductTVector(map(similar, x.parts)...)
+similar(x::ProductTVector, ::Type{T}) where T = ProductTVector(map(t -> similar(t, T), x.parts)...)
+
+(+)(v1::ProductTVector, v2::ProductTVector) = ProductTVector(map(+, v1.parts, v2.parts)...)
+(-)(v1::ProductTVector, v2::ProductTVector) = ProductTVector(map(-, v1.parts, v2.parts)...)
+(-)(v::ProductTVector) = ProductTVector(map(-, v.parts))
+(*)(a::Number, v::ProductTVector) = ProductTVector(map(t -> a*t, v.parts))
+
+function proj_product(x::ProductTVector, i::Integer)
+    return x.parts[i]
+end
+
+"""
+    ProductCoTVector(parts)
+
+A more general but slower representation of cotangent vectors on a product
+manifold.
+"""
+struct ProductCoTVector{TM<:Tuple} <: CoTVector
+    parts::TM
+end
+
+ProductCoTVector(covectors...) = ProductCoTVector{typeof(covectors)}(covectors)
+eltype(x::ProductCoTVector) = eltype(Tuple{map(eltype, x.parts)...})
+similar(x::ProductCoTVector) = ProductCoTVector(map(similar, x.parts)...)
+similar(x::ProductCoTVector, ::Type{T}) where T = ProductCoTVector(map(t -> similar(t, T), x.parts)...)
+
+function proj_product(x::ProductCoTVector, i::Integer)
+    return x.parts[i]
+end
+
+function isapprox(M::ProductManifold, x, y; kwargs...)
+    return all(t -> isapprox(t...; kwargs...), ziptuples(M.manifolds, x.parts, y.parts))
+end
+
+function isapprox(M::ProductManifold, x, v, w; kwargs...)
+    return all(t -> isapprox(t...; kwargs...), ziptuples(M.manifolds, x.parts, v.parts, w.parts))
+end
+
+function representation_size(M::ProductManifold, ::Type{T}) where {T}
+    return (mapreduce(m -> sum(representation_size(m, T)), +, M.manifolds),)
+end
+
+manifold_dimension(M::ProductManifold) = sum(map(m -> manifold_dimension(m), M.manifolds))
+
+struct ProductMetric <: Metric end
+
+function det_local_metric(M::MetricManifold{ProductManifold, ProductMetric}, x::ProductArray)
+    dets = map(det_local_metric, M.manifolds, x.parts)
+    return prod(dets)
+end
+
+function inner(M::ProductManifold, x, v, w)
+    subproducts = map(inner, M.manifolds, x.parts, v.parts, w.parts)
+    return sum(subproducts)
+end
+
+function exp!(M::ProductManifold, y, x, v)
+    map(exp!, M.manifolds, y.parts, x.parts, v.parts)
+    return y
+end
+
+function exp(M::ProductManifold, x::ProductMPoint, v::ProductTVector)
+    return ProductMPoint(map(exp, M.manifolds, x.parts, v.parts)...)
+end
+
+function log!(M::ProductManifold, v, x, y)
+    map(log!, M.manifolds, v.parts, x.parts, y.parts)
+    return v
+end
+
+function log(M::ProductManifold, x::ProductMPoint, y::ProductMPoint)
+    return ProductTVector(map(log, M.manifolds, x.parts, y.parts)...)
+end
+
+function distance(M::ProductManifold, x, y)
+    return sqrt(sum(map(distance, M.manifolds, x.parts, y.parts).^2))
```

Comment:
Can use `mapreduce`. Between `map`, `sum`, and `.^2`, it looks like the manifolds are traversed 3 times, when it could be done in one.

Reply:
Also, this should probably be further up with `inner`, since that's where we usually put it.

Reply:
I moved it up but again, `mapreduce` doesn't iterate over multiple collections.

Reply:
Couldn't you `zip` them?
```julia
sqrt(mapreduce(u -> distance(u...)^2, +, zip(M.manifolds, x.parts, y.parts)))
```
Not sure it's worth all the effort, though...

Reply:
I could but see here: https://github.com/JuliaNLSolvers/Manifolds.jl/pull/22#discussion_r300322315